src/bin/scheduler/build.rs

```diff
@@ -63,30 +63,82 @@ fn build_environment(
     };
     let base_path = &working_directory.join(session.id()).join(id);
     info!("Building environment for plan {id}");
-    let run_spec = RunSpec {
-        id: &format!("robotmk_env_building_{id}"),
-        command_spec: &build_instructions.command_spec,
-        base_path,
-        timeout: build_instructions.timeout,
-        cancellation_token,
-    };
     let start_time = Utc::now();
     build_stage_reporter.update(
         id,
         EnvironmentBuildStage::InProgress(start_time.timestamp()),
     )?;
-    let outcome = run_build_command(id, &run_spec, session, start_time)?;
+    let outcome = run_build_commands(
+        id,
+        &build_instructions,
+        session,
+        start_time,
+        cancellation_token,
+        base_path,
+    )?;
     build_stage_reporter.update(id, EnvironmentBuildStage::Complete(outcome.clone()))?;
     Ok(outcome)
 }
 
-fn run_build_command(
+fn run_build_commands(
     id: &str,
-    run_spec: &RunSpec,
+    build_instructions: &BuildInstructions,
     session: &Session,
     start_time: DateTime<Utc>,
+    cancellation_token: &CancellationToken,
+    base_path: &Utf8Path,
 ) -> Result<BuildOutcome, Cancelled> {
-    match session.run(&run_spec) {
+    if let Some(command_spec) = &build_instructions.import_command_spec {
+        let import_run_spec = RunSpec {
+            id: &format!("robotmk_env_import_{id}"),
+            command_spec,
+            base_path,
+            timeout: build_instructions.timeout,
+            cancellation_token,
+        };
+        match session.run(&import_run_spec) {
+            Ok(Outcome::Completed(0)) => {
+                info!("Environment import succeeded for plan {id}");
+            }
+            Ok(Outcome::Completed(_exit_code)) => {
+                error!("Environment import not successful, plan {id} will be dropped");
+                return Ok(BuildOutcome::Error(format!(
+                    "Environment import not successful, see {base_path} for stdio logs"
+                )));
+            }
+            Ok(Outcome::Timeout) => {
+                error!("Environment import timed out, plan {id} will be dropped");
+                return Ok(BuildOutcome::Timeout);
+            }
+            Ok(Outcome::Cancel) => {
+                error!("Environment import cancelled, plan {id} will be dropped");
+                return Err(Cancelled {});
+            }
+            Err(e) => {
+                let log_error = e.context(anyhow!(
+                    "Environment import failed, plan {id} will be dropped. See {} for stdio logs",
+                    base_path,
+                ));
+                error!("{log_error:?}");
+                return Ok(BuildOutcome::Error(format!("{log_error:?}")));
+            }
```

Comment:
Would it make sense to outsource this into a separate function to reduce code duplication?

Reply:
I think the similiarity here is just coincidental. For example, if ever need to handle the exit codes, then this would need to be de-duplicated. I usually like to:
* only abstract once there are three or more examples,
* don't code share messaging/ui elements (in this case the log messages).

We can deviate from my gut feeling here, but the duplication here was deliberate. Adapting duplicated code is much easier, then adapting code with unnecessary coupling.

Reply:
Sounds reasonable to me, thanks for clarifying.